File: sdks/java/src/main/java/com/norikv/client/internal/router/Router.java
package com.norikv.client.internal.router;

import com.norikv.client.hash.HashFunctions;
import com.norikv.client.internal.conn.ConnectionPool;
import com.norikv.client.types.ConnectionException;
import com.norikv.client.types.NotLeaderException;
import io.grpc.ManagedChannel;

import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

/**
 * Routes requests to the correct shard leader with caching.
 *
 * <p>This class provides smart routing by:
 * <ul>
 * <li>Computing shard assignments using consistent hashing</li>
 * <li>Caching leader information for each shard</li>
 * <li>Falling back to random node on cache miss</li>
 * <li>Updating cache on NotLeaderException hints</li>
 * </ul>
 *
 * <p>Thread-safe for concurrent use.
 */
public final class Router {
    private final List<String> nodes;
    private final int totalShards;
    private final ConnectionPool connectionPool;

    // Leader cache: shard_id -> node_address
    private final ConcurrentHashMap<Integer, String> leaderCache;

    // Lock for coordinating cache updates
    private final ReadWriteLock cacheLock;

    // Round-robin index for fallback node selection
    private int fallbackIndex;

    /**
     * Creates a new Router.
     *
     * @param nodes list of cluster node addresses
     * @param totalShards total number of shards in the cluster
     * @param connectionPool connection pool for gRPC channels
     */
    public Router(List<String> nodes, int totalShards, ConnectionPool connectionPool) {
        if (nodes == null || nodes.isEmpty()) {
            throw new IllegalArgumentException("nodes cannot be null or empty");
        }
        if (totalShards <= 0) {
            throw new IllegalArgumentException("totalShards must be > 0");
        }
        if (connectionPool == null) {
            throw new IllegalArgumentException("connectionPool cannot be null");
        }

        this.nodes = nodes;
        this.totalShards = totalShards;
        this.connectionPool = connectionPool;
        this.leaderCache = new ConcurrentHashMap<>();
        this.cacheLock = new ReentrantReadWriteLock();
        this.fallbackIndex = 0;
    }

    /**
     * Gets the shard ID for a given key.
     *
     * <p>Uses xxhash64 + jump consistent hash to deterministically map the key to a shard.
     *
     * @param key the key
     * @return the shard ID (0-based)
     */
    public int getShardForKey(byte[] key) {
        return HashFunctions.getShardForKey(key, totalShards);
    }

    /**
     * Gets a channel to the leader for the given shard.
     *
     * <p>Attempts to use cached leader information. If no leader is cached,
     * falls back to a random node from the cluster.
     *
     * @param shardId the shard ID
     * @return routing info containing channel and whether leader was cached
     * @throws ConnectionException if unable to get channel
     */
    public RouteInfo getChannelForShard(int shardId) throws ConnectionException {
        // Try cached leader first
        String cachedLeader = leaderCache.get(shardId);
        if (cachedLeader != null) {
            ManagedChannel channel = connectionPool.getChannel(cachedLeader);
            return new RouteInfo(channel, cachedLeader, shardId, true);
        }

        // No cached leader, use fallback node
        String fallbackNode = getFallbackNode();
        ManagedChannel channel = connectionPool.getChannel(fallbackNode);
        return new RouteInfo(channel, fallbackNode, shardId, false);
    }

    /**
     * Gets a channel to the leader for the given key.
     *
     * <p>First computes the shard ID from the key, then gets the channel for that shard.
     *
     * @param key the key
     * @return routing info containing channel and shard information
     * @throws ConnectionException if unable to get channel
     */
    public RouteInfo getChannelForKey(byte[] key) throws ConnectionException {
        int shardId = getShardForKey(key);
        return getChannelForShard(shardId);
    }

    /**
     * Updates the cached leader for a shard.
     *
     * <p>Called when receiving NotLeaderException with a leader hint.
     *
     * @param shardId the shard ID
     * @param leaderAddress the leader node address
     */
    public void updateLeader(int shardId, String leaderAddress) {
        if (leaderAddress == null || leaderAddress.isEmpty()) {
            return;
        }

        cacheLock.writeLock().lock();
        try {
            leaderCache.put(shardId, leaderAddress);
        } finally {
            cacheLock.writeLock().unlock();
        }
    }

    /**
     * Handles a NotLeaderException by updating the leader cache.
     *
     * <p>Extracts the leader hint from the exception and updates the cache.
     *
     * @param error the NotLeaderException
     */
    public void handleNotLeader(NotLeaderException error) {
        String leaderHint = error.getLeaderHint();
        int shardId = error.getShardId();

        if (leaderHint != null && !leaderHint.isEmpty()) {
            updateLeader(shardId, leaderHint);
        } else {
            // No hint provided, invalidate cache entry
            invalidateLeader(shardId);
        }
    }

    /**
     * Invalidates the cached leader for a shard.
     *
     * <p>Called when we receive NotLeaderException without a hint,
     * or when we suspect the cached leader is stale.
     *
     * @param shardId the shard ID
     */
    public void invalidateLeader(int shardId) {
        cacheLock.writeLock().lock();
        try {
            leaderCache.remove(shardId);
        } finally {
            cacheLock.writeLock().unlock();
        }
    }

    /**
     * Invalidates all cached leader information.
     *
     * <p>Useful when the cluster topology changes significantly.
     */
    public void invalidateAllLeaders() {
        cacheLock.writeLock().lock();
        try {
            leaderCache.clear();
        } finally {
            cacheLock.writeLock().unlock();
        }
    }

    /**
     * Gets the cached leader for a shard, if any.
     *
     * @param shardId the shard ID
     * @return the leader address, or null if not cached
     */
    public String getCachedLeader(int shardId) {
        return leaderCache.get(shardId);
    }

    /**
     * Checks if a leader is cached for the given shard.
     *
     * @param shardId the shard ID
     * @return true if a leader is cached
     */
    public boolean hasLeaderCached(int shardId) {
        return leaderCache.containsKey(shardId);
    }

    /**
     * Gets the number of cached leaders.
     *
     * @return the number of cached leaders
     */
    public int getCachedLeaderCount() {
        return leaderCache.size();
    }

    /**
     * Gets a fallback node using round-robin selection.
     *
     * <p>Used when no leader is cached for a shard.
     *
     * @return a node address
     */
    private synchronized String getFallbackNode() {
        String node = nodes.get(fallbackIndex % nodes.size());
        fallbackIndex++;
        return node;
    }

    /**
     * Gets the list of cluster nodes.
     *
     * @return the node addresses
     */
    public List<String> getNodes() {
        return nodes;
    }

    /**
     * Gets the total number of shards.
     *
     * @return the total shards
     */
    public int getTotalShards() {
        return totalShards;
    }

    /**
     * Gets router statistics.
     *
     * @return router statistics
     */
    public RouterStats getStats() {
        return new RouterStats(
                nodes.size(),
                totalShards,
                leaderCache.size()
        );
    }

    /**
     * Information about a route to a shard.
     */
    public static final class RouteInfo {
        private final ManagedChannel channel;
        private final String nodeAddress;
        private final int shardId;
        private final boolean leaderCached;

        RouteInfo(ManagedChannel channel, String nodeAddress, int shardId, boolean leaderCached) {
            this.channel = channel;
            this.nodeAddress = nodeAddress;
            this.shardId = shardId;
            this.leaderCached = leaderCached;
        }

        /**
         * Gets the gRPC channel for this route.
         *
         * @return the managed channel
         */
        public ManagedChannel getChannel() {
            return channel;
        }

        /**
         * Gets the node address for this route.
         *
         * @return the node address
         */
        public String getNodeAddress() {
            return nodeAddress;
        }

        /**
         * Gets the shard ID for this route.
         *
         * @return the shard ID
         */
        public int getShardId() {
            return shardId;
        }

        /**
         * Checks if the leader was cached.
         *
         * @return true if leader was cached, false if using fallback
         */
        public boolean isLeaderCached() {
            return leaderCached;
        }

        @Override
        public String toString() {
            return "RouteInfo{" +
                    "nodeAddress='" + nodeAddress + '\'' +
                    ", shardId=" + shardId +
                    ", leaderCached=" + leaderCached +
                    '}';
        }
    }

    /**
     * Statistics about the router.
     */
    public static final class RouterStats {
        private final int totalNodes;
        private final int totalShards;
        private final int cachedLeaders;

        RouterStats(int totalNodes, int totalShards, int cachedLeaders) {
            this.totalNodes = totalNodes;
            this.totalShards = totalShards;
            this.cachedLeaders = cachedLeaders;
        }

        public int getTotalNodes() {
            return totalNodes;
        }

        public int getTotalShards() {
            return totalShards;
        }

        public int getCachedLeaders() {
            return cachedLeaders;
        }

        /**
         * Gets the cache hit ratio (cached leaders / total shards).
         *
         * @return cache hit ratio (0.0 to 1.0)
         */
        public double getCacheHitRatio() {
            return totalShards > 0 ? (double) cachedLeaders / totalShards : 0.0;
        }

        @Override
        public String toString() {
            return "RouterStats{" +
                    "totalNodes=" + totalNodes +
                    ", totalShards=" + totalShards +
                    ", cachedLeaders=" + cachedLeaders +
                    ", cacheHitRatio=" + String.format("%.2f", getCacheHitRatio()) +
                    '}';
        }
    }
}

File: sdks/go/internal/router/router.go
// Package router implements client-side routing for NoriKV.
package router

import (
	"context"
	"fmt"
	"sync"

	"github.com/norikv/norikv-go/hash"
	"golang.org/x/sync/singleflight"
	"google.golang.org/grpc"
	"google.golang.org/grpc/metadata"

	"github.com/norikv/norikv-go/internal/conn"
)

// Router handles routing requests to the appropriate node based on key hashing
// and leader information. Uses single-flight pattern to deduplicate concurrent
// leader lookups for the same shard.
type Router struct {
	pool        *conn.Pool
	totalShards int

	// Leader tracking
	mu            sync.RWMutex
	leaderCache   map[uint32]string // shard ID -> leader address
	nodeAddresses []string          // all known node addresses

	// Single-flight for leader discovery
	leaderLookup singleflight.Group
}

// NewRouter creates a new router with the given connection pool and configuration.
func NewRouter(pool *conn.Pool, totalShards int, nodeAddresses []string) *Router {
	return &Router{
		pool:          pool,
		totalShards:   totalShards,
		leaderCache:   make(map[uint32]string),
		nodeAddresses: nodeAddresses,
	}
}

// GetShardForKey returns the shard ID for the given key.
func (r *Router) GetShardForKey(key []byte) uint32 {
	return uint32(hash.GetShardForKey(key, r.totalShards))
}

// GetLeader returns the cached leader address for the given shard.
// Returns empty string if no leader is cached.
func (r *Router) GetLeader(shardID uint32) string {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.leaderCache[shardID]
}

// SetLeader updates the cached leader address for the given shard.
func (r *Router) SetLeader(shardID uint32, address string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.leaderCache[shardID] = address
}

// ClearLeader removes the cached leader for the given shard.
// Used when receiving NOT_LEADER errors.
func (r *Router) ClearLeader(shardID uint32) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.leaderCache, shardID)
}

// UpdateNodeAddresses updates the list of known node addresses.
// Called by the topology manager when cluster membership changes.
func (r *Router) UpdateNodeAddresses(addresses []string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.nodeAddresses = addresses
}

// GetConnection returns a gRPC connection for the given shard.
// First tries to use the cached leader, falls back to any node if no leader is cached.
// Uses single-flight pattern to deduplicate concurrent lookups for the same shard.
func (r *Router) GetConnection(ctx context.Context, shardID uint32) (*grpc.ClientConn, string, error) {
	// Fast path: try cached leader
	leader := r.GetLeader(shardID)
	if leader != "" {
		conn, err := r.pool.Get(ctx, leader)
		if err == nil {
			return conn, leader, nil
		}
		// Leader connection failed, clear cache and continue
		r.ClearLeader(shardID)
	}

	// Slow path: discover leader using single-flight
	// This ensures only one goroutine performs leader discovery per shard
	key := fmt.Sprintf("shard:%d", shardID)
	result, err, _ := r.leaderLookup.Do(key, func() (interface{}, error) {
		return r.discoverLeader(ctx, shardID)
	})

	if err != nil {
		return nil, "", err
	}

	address := result.(string)
	conn, err := r.pool.Get(ctx, address)
	if err != nil {
		return nil, "", fmt.Errorf("failed to connect to %s: %w", address, err)
	}

	return conn, address, nil
}

// discoverLeader attempts to discover the leader for the given shard.
// Tries all known nodes until one responds successfully or returns leader hint.
func (r *Router) discoverLeader(ctx context.Context, shardID uint32) (string, error) {
	r.mu.RLock()
	nodes := make([]string, len(r.nodeAddresses))
	copy(nodes, r.nodeAddresses)
	r.mu.RUnlock()

	if len(nodes) == 0 {
		return "", fmt.Errorf("no nodes available for shard %d", shardID)
	}

	// Try each node until we find the leader or get a leader hint
	for _, address := range nodes {
		_, err := r.pool.Get(ctx, address)
		if err != nil {
			// Connection failed, try next node
			continue
		}

		// Successfully connected - cache this node as potential leader
		r.SetLeader(shardID, address)
		return address, nil
	}

	return "", fmt.Errorf("failed to discover leader for shard %d: all nodes unreachable", shardID)
}

// GetConnectionForKey returns a gRPC connection for the node responsible for the given key.
// Convenience method that combines GetShardForKey and GetConnection.
func (r *Router) GetConnectionForKey(ctx context.Context, key []byte) (*grpc.ClientConn, string, uint32, error) {
	shardID := r.GetShardForKey(key)
	conn, address, err := r.GetConnection(ctx, shardID)
	if err != nil {
		return nil, "", shardID, err
	}
	return conn, address, shardID, nil
}

// HandleNotLeaderError processes a NOT_LEADER error by updating the leader cache.
// Extracts the leader hint from the error/metadata and updates the cache.
func (r *Router) HandleNotLeaderError(shardID uint32, leaderHint string) {
	// Clear old leader
	r.ClearLeader(shardID)

	// If we got a leader hint, cache it
	if leaderHint != "" {
		r.SetLeader(shardID, leaderHint)
	}
}

// ExtractLeaderHint extracts the leader hint from gRPC metadata.
// Returns empty string if no hint is present.
func ExtractLeaderHint(md metadata.MD) string {
	if md == nil {
		return ""
	}
	hints := md.Get("leader-hint")
	if len(hints) > 0 {
		return hints[0]
	}
	return ""
}

// GetAllLeaders returns a copy of the current leader cache.
// Useful for debugging and monitoring.
func (r *Router) GetAllLeaders() map[uint32]string {
	r.mu.RLock()
	defer r.mu.RUnlock()

	leaders := make(map[uint32]string, len(r.leaderCache))
	for shardID, address := range r.leaderCache {
		leaders[shardID] = address
	}
	return leaders
}

// ClearAllLeaders removes all cached leader information.
// Used when cluster topology changes significantly.
func (r *Router) ClearAllLeaders() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.leaderCache = make(map[uint32]string)
}

// GetNodeAddresses returns the current list of known node addresses.
func (r *Router) GetNodeAddresses() []string {
	r.mu.RLock()
	defer r.mu.RUnlock()

	addresses := make([]string, len(r.nodeAddresses))
	copy(addresses, r.nodeAddresses)
	return addresses
}

// TotalShards returns the total number of shards.
func (r *Router) TotalShards() int {
	return r.totalShards
}

// Close closes the router and cleans up resources.
func (r *Router) Close() error {
	r.ClearAllLeaders()
	return nil
}
